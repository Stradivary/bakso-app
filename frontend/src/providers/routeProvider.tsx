import { createBrowserRouter, RouterProvider } from "react-router-dom";
import { ProtectedRoute } from "../components/ProtectedRoute";
import { Register, MainPage, CustomerPage } from "../pages";

const router = createBrowserRouter([
  {
    path: '/auth',
    element: <Register />,
  },
  {
    path: '/',
    element: (
      <ProtectedRoute>
        <CustomerPage />
      </ProtectedRoute>
    ),
  },
  {
    path: '/maps',
    element: <CustomerPage />,
  },

]);


export function RouteProvider() {
  return (
    <RouterProvider router={router}  />
  );
}