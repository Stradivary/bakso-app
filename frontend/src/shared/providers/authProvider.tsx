import { PropsWithChildren, useState, useEffect } from "react";
import { User } from "../models/user";
import { AuthContext } from "../services/authService";


export function AuthProvider({ children }: Readonly<PropsWithChildren>) {
  const [user, setUser] = useState<User | null>(null);

  useEffect(() => {
    const storedUser = localStorage.getItem('user');
    if (storedUser) {
      setUser(JSON.parse(storedUser));
    }
  }, []);

  const login = (userData: User) => {
    setUser(userData);
    localStorage.setItem('user', JSON.stringify(userData));
  };

  const logout = () => {
    setUser(null);
    localStorage.removeItem('user');
  };

  return (
    <AuthContext.Provider value={{ user, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
}
