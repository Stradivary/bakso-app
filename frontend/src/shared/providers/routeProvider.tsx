import { createBrowserRouter, RouterProvider } from "react-router-dom";
import { ProtectedRoute } from "../components/ProtectedRoute";
import { AuthPage, CustomerPage } from "../../pages";

const router = createBrowserRouter([
  {
    path: '/auth',
    element: <AuthPage />,
  },
  {
    path: '/',
    element: (
      <ProtectedRoute>
        <CustomerPage />
      </ProtectedRoute>
    ),
  },
]);


export function RouteProvider() {
  return (
    <RouterProvider router={router} />
  );
}