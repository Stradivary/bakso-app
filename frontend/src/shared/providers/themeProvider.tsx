import { MantineProvider } from "@mantine/core";
import { ModalsProvider } from "@mantine/modals";
import { Notifications } from "@mantine/notifications";
import { PropsWithChildren } from "react";

export const ThemeProvider = ({ children }: PropsWithChildren) => {
  return (
    <MantineProvider theme={{ primaryColor: "red" }}>
      <ModalsProvider>
        {children}
        <Notifications position="top-center" />
      </ModalsProvider>
    </MantineProvider>
  );
};
