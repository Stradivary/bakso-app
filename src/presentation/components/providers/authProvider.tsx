import { Database } from "@/data/models/supabase.types";
import { signInUser } from "@/domain/services/authService";
import { supabase } from "@/domain/services/supabaseService";
import { notifications } from "@mantine/notifications";
import { Session } from "@supabase/supabase-js";
import {
  createContext,
  PropsWithChildren,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

type UserDetails = Database["public"]["Tables"]["user_profiles"]["Row"];

export interface AuthContextType {
  session: Session | null;
  user: UserDetails | null;
  isLoading: boolean;
  error: string | null;
  login: (
    name: string,
    role: string,
    latitude: number,
    longitude: number,
  ) => Promise<void>;
  logout: () => Promise<void>;
  isAuthenticated: boolean;
}

export const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [session, setSession] = useState<Session | null>(null);
  const [user, setUser] = useState<UserDetails | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const [initialized, setInitialized] = useState(false);

  // Fetch user profile data
  const fetchUserProfile = useCallback(async (userId: string) => {
    try {
      const { data, error: fetchError } = await supabase
        .from("user_profiles")
        .select("*")
        .eq("id", userId)
        .single();

      if (fetchError) throw fetchError;

      if (data) {
        setUser(data);
      }
    } catch (error) {
      console.error("Error fetching user profile:", error);
      setError("Failed to fetch user profile");
    }
  }, []);

  // Initialize auth state
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        // Check for existing session
        const {
          data: { session: existingSession },
          error: sessionError,
        } = await supabase.auth.getSession();

        if (sessionError) {
          console.error("failed to set session, log in again");
          throw sessionError;
        }

        if (existingSession) {
          setSession(existingSession);
          await fetchUserProfile(existingSession.user.id);
        }
      } catch (error) {
        console.error("Error initializing auth:", error);
        setError("Failed to initialize authentication");
      } finally {
        setIsLoading(false);
        setInitialized(true);
      }
    };

    initializeAuth();
  }, [fetchUserProfile]);
  const login = useCallback(
    async (name: string, role: string, latitude: number, longitude: number) => {
     
      try {
        setIsLoading(true);
        const { session: newSession, error: signInError } = await signInUser(
          name,
          role,
          latitude,
          longitude,
        );
        if (signInError) throw signInError;

        setSession(newSession);
        if (newSession) {
          await fetchUserProfile(newSession.user.id);
        }
        setError(null);
      } catch (error) {
        console.error("Error during login:", error);
        setError("Failed to login");
      } finally {
        setIsLoading(false);
      }
    },
    [fetchUserProfile],
  );

  const logout = useCallback(async () => {
    try {
      const { error: signOutError } = await supabase.auth.signOut({
        scope: "local",
      });

      if (signOutError) throw signOutError;
      setSession(null);
      setUser(null);
    } catch (error) {
      console.error("Error during logout:", error);
      setError("Failed to logout");
    }
  }, []);

  // Storage change observer
  useEffect(() => {
    if (!initialized) return;

    const handleStorageChange = (event: StorageEvent) => {
      const currentSession = event.newValue;
      const previousSession = event.oldValue;

      if (event.key === "abangbakso-session") {
        if (!previousSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session injection detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
          return;
        }

        if (!currentSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session removal detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
          return;
        }
        // If session was changed externally
        if (currentSession !== previousSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session tampering detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
        }
      }
    };

    // Add storage event listener
    window.addEventListener("storage", handleStorageChange);

    return () => {
      window.removeEventListener("storage", handleStorageChange);
    };
  }, [initialized, logout, session?.user.id]);

  useEffect(() => {
    if (!initialized) return;

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange(async (event, newSession) => { 

      if (!newSession?.user.id) return;
      if ((newSession?.user.id && newSession?.user.id) && newSession?.user.id !== session?.user.id) {
        return;
      }
      switch (event) {
        case "INITIAL_SESSION":
          break;
        case "SIGNED_IN":
        case "TOKEN_REFRESHED":
        case "USER_UPDATED":
          setSession(newSession);
          if (newSession) {
            await fetchUserProfile(newSession.user.id);
          }
          break;
        case "SIGNED_OUT":
          setSession(null);
          setUser(null);
          break;
      }
    });

    return () => {
      subscription.unsubscribe();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialized, fetchUserProfile]);

  const isAuthenticated = !!session?.user?.id && !isLoading;

  const values = useMemo(
    () => ({
      session,
      user,
      isLoading,
      error,
      login,
      logout,
      isAuthenticated,
    }),
    [session, user, isLoading, error, login, logout, isAuthenticated],
  );

  return <AuthContext.Provider value={values}>{children}</AuthContext.Provider>;
};

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined,
);

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
