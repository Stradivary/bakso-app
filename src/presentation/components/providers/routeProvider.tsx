import { Container } from "@mantine/core";
import {
  createBrowserRouter,
  Outlet,
  RouteObject,
  RouterProvider,
} from "react-router-dom";
import { OfflineHandler } from "../OfflineHandler";
import { BroadcastMapPage } from "@/presentation/views/map";
import LoginView from "@/presentation/views/login/LoginView";
const routes: RouteObject[] = [
  {
    path: "/",
    element: (
      <>
        <Outlet />
        <OfflineHandler />
      </>
    ),
    children: [
      {
        index: true,
        element: <BroadcastMapPage />,
      },
      {
        path: "/login",
        element: <LoginView />,
      },
      {
        path: "*",
        element: (
          <Container h="100dvh">
            <h1>Oops, halaman tidak ditemukan</h1>

            <p>
              <a href="/">Kembali ke halaman utama</a>
            </p>
          </Container>
        ),
      },
    ] satisfies RouteObject[],
  },
] satisfies RouteObject[];

const router = createBrowserRouter(routes, {
  future: {
    v7_relativeSplatPath: true,
  },
});

export function RouteProvider() {
  return <RouterProvider router={router} />;
}
