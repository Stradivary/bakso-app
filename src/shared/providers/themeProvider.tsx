import { createTheme, MantineColorsTuple, MantineProvider } from "@mantine/core";
import { ModalsProvider } from "@mantine/modals";
import { Notifications } from "@mantine/notifications";
import { PropsWithChildren } from "react";

const primary: MantineColorsTuple = [
  '#ffe8ed',
  '#ffcfd7',
  '#fe9cab',
  '#fd677d',
  '#fc3a56',
  '#fb203e',
  '#fc1131',
  '#e10224',
  '#c9001f',
  '#b00018'
];

const secondary: MantineColorsTuple = [
  '#ebf4ff',
  '#d4e5fa',
  '#a3c9f7',
  '#71abf6',
  '#4b92f5',
  '#3782f5',
  '#2e7af6',
  '#2368dc',
  '#195dc5',
  '#0050ad'
];

const theme = createTheme({

  colors: {
    primary,
    secondary,
  },
  primaryColor: 'primary',
  defaultRadius: 8,
  fontFamily: "'Poppins', system-ui, sans-serif",
  headings: {
    fontFamily: "'Telkomsel Batik Sans', system-ui, sans-serif",
  }
});

export const ThemeProvider = ({ children }: PropsWithChildren) => {
  return (
    <MantineProvider theme={theme}>
      <ModalsProvider>
        {children}
        <Notifications position="top-center" />
      </ModalsProvider>
    </MantineProvider>
  );
};
