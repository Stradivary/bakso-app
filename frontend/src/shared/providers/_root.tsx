import { MantineProvider } from "@mantine/core";
import { Notifications } from "@mantine/notifications";
import { AuthProvider } from "./authProvider";

import "@mantine/core/styles.css";
import "@mantine/notifications/styles.css";
import { RouteProvider } from "./routeProvider";


export function Root() {
  return (
    <AuthProvider>
      <MantineProvider theme={{ primaryColor: "red" }}>
        <RouteProvider />
        <Notifications />
      </MantineProvider>
    </AuthProvider>
  );
}

