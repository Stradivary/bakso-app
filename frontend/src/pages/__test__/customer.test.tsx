import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import LocationMapPage from "../customer";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "../../shared/providers/themeProvider";

describe("LocationMapPage", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ThemeProvider>
          <LocationMapPage />
        </ThemeProvider>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
