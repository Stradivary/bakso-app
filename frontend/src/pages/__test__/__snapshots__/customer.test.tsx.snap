// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LocationMapPage > should match snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-primary-color-filled: var(--mantine-color-red-filled);--mantine-primary-color-filled-hover: var(--mantine-color-red-filled-hover);--mantine-primary-color-light: var(--mantine-color-red-light);--mantine-primary-color-light-hover: var(--mantine-color-red-light-hover);--mantine-primary-color-light-color: var(--mantine-color-red-light-color);--mantine-primary-color-0: var(--mantine-color-red-0);--mantine-primary-color-1: var(--mantine-color-red-1);--mantine-primary-color-2: var(--mantine-color-red-2);--mantine-primary-color-3: var(--mantine-color-red-3);--mantine-primary-color-4: var(--mantine-color-red-4);--mantine-primary-color-5: var(--mantine-color-red-5);--mantine-primary-color-6: var(--mantine-color-red-6);--mantine-primary-color-7: var(--mantine-color-red-7);--mantine-primary-color-8: var(--mantine-color-red-8);--mantine-primary-color-9: var(--mantine-color-red-9);}:root[data-mantine-color-scheme="dark"]{--mantine-color-anchor: var(--mantine-color-red-4);}:root[data-mantine-color-scheme="light"]{--mantine-color-anchor: var(--mantine-color-red-6);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_4451eb3a mantine-Center-root"
  >
    Memuat Lokasi...
  </div>
</DocumentFragment>
`;
